### PR TITLE
upgrade prisma and prisma client

### PR DESCRIPTION
**Pull request by Flip-Liquid**
## Linear Ticket

[Pro-322](https://linear.app/govrn/issue/PRO-332/upgrade-to-prisma-4)

## Description

Upgrade prisma and prisma client to v 4

## Screencaptures

For frontend changes, considering adding a screenshot or screencapture of the changes.
Before/after screenshots may be helpful when relevant.

### CODE REVIEW

**Comment by Flip-Liquid** [comment]
@alexkeating src/prisma/resolvers/Contribution.ts L652 is the only area potentially affected by the upgrade. blocked on testing it because of the DB permissions issue

**Comment by alexkeating** [comment]
Okay sounds good, let me know when tested and looks good. I can do a once over and test as well

**Comment by Flip-Liquid** [comment]
@jonathanprozzi mind giving the app a quick spin with this?

File: apps/protocol-api/src/main.ts
Comment (alexkeating) [inline apps/protocol-api/src/main.ts:124]: Damn this breaks our naming convention

File: package.json
Comment (alexkeating) [inline package.json:47]: We should  probably be more specific here and have "^4.7.1"

File: package.json
Comment (alexkeating) [inline package.json:185]: This is a duplicate and can be removed since we have prisma in dependencies

Comment (Flip-Liquid) [reply package.json:185]: prisma is for the CLI and @prisma/client is the module where the autogenerated client actually sits. I don't think these are duplicates. see https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client

File: apps/protocol-api/src/prisma/resolvers/Contribution.ts
Comment (Flip-Liquid) [inline apps/protocol-api/src/prisma/resolvers/Contribution.ts:791]: this is bad and i feel bad

**Comment by jonathanprozzi** [comment]
Adding notes from our testing session last night:
- New types from Prisma 4 caused issues with the frontend, particularly the Dashboards and `getContributionCountByDateForUserInRange`
- Type for `count` was changed to a `number` and not a `string` which was the primary type issue
- Queries are working, need to make some slight fixes to the Heat Map charts since they expect the date in a different format -- will address this in a separate PR after this is merged

To do:
- Build is breaking because of type differences for the Dates 
- After merging in, we will address the date issue for the heat maps in an immediate PR (basically, the heat map expects dates in a very particular format (with no time value) and so we have to do a small conversion prior to passing into the heat map. Already tested and it works, but don't want to add it to this PR

@Flip-Liquid I know I omitted a lot of the details from last night -- you crushed the quick fixes!

**Comment by jonathanprozzi** [comment]
That resolved the build errors from the type, but will look into Cypress run

**Comment by Flip-Liquid** [comment]
thanks for the help on the FE changes @jonathanprozzi

**Comment by jonathanprozzi** [comment]
> thanks for the help on the FE changes @jonathanprozzi

For sure! super close, not sure why the Cypress run is failing but I think after we resolve that we can get this merged in!

**Comment by alexkeating** [comment]
> Adding notes from our testing session last night:
> 
>     * New types from Prisma 4 caused issues with the frontend, particularly the Dashboards and `getContributionCountByDateForUserInRange`
> 
>     * Type for `count` was changed to a `number` and not a `string` which was the primary type issue
> 
>     * Queries are working, need to make some slight fixes to the Heat Map charts since they expect the date in a different format -- will address this in a separate PR after this is merged
> 
> 
> To do:
> 
>     * Build is breaking because of type differences for the Dates
> 
>     * After merging in, we will address the date issue for the heat maps in an immediate PR (basically, the heat map expects dates in a very particular format (with no time value) and so we have to do a small conversion prior to passing into the heat map. Already tested and it works, but don't want to add it to this PR
> 
> 
> @Flip-Liquid I know I omitted a lot of the details from last night -- you crushed the quick fixes!
> 
> Update:
> 
>     * I added the heatmap fixes since I pushed something up to resolve the type changes preventing the build. Since I already touched those files any way to resolve the build issue I left the format changes in

+1 Thanks for fixing those before merging

**Comment by Flip-Liquid** [comment]
@jonathanprozzi lets merge before #331, I need to send an iteration on there anyway

**Comment by alexkeating** [comment]
I tried applying the migrations locally and I am getting this error?

Screenshot from 2022-12-08 08-06-15

**Comment by Flip-Liquid** [comment]
That looks like it's when you're running the protocol-api right? Did you
run migrate/generate/generate-gql

**Comment by alexkeating** [comment]
> migrate/generate/generate-gql

Yes, that was with protocol-api. I ran `nx run protocol-api:migrate`. Should

> generate-gql
> [](#)

Ah I think I found the issue. I forgot to run yarn install

**Comment by Flip-Liquid** [comment]
@alexkeating looks like we're good to go here pending your review 💯

**Review by alexkeating — APPROVED** [review]
Looks good to me

**Pull request merged**
